File: 백준/Gold/2293. 동전 1/동전 1.cc
#include <iostream>

using namespace std;

const int MX = 100002;
int coins[102];
int dp[MX];

int main(){
    int n, k;
    cin >> n >> k;
    
    for(int i = 0; i < n; i++){
        cin >> coins[i];
    }
    dp[0] = 1;
    for(int i = 0; i < n; i++){
        int coin = coins[i];
        for(int j = 1; j <= 10000; j++){
            if(j-coin <0) continue;
            dp[j] += dp[j-coin];
        }
    }
    
    cout << dp[k];

    return 0;
}

File: 백준/Gold/2293. 동전 1/동전 1.java
import java.io.*;
import java.util.*;

public class Main {


    public static void main(String[] args) throws Exception {
        // System.setIn(new FileInputStream("src/input.txt"));
        Scanner sc = new Scanner(System.in);

        int N = sc.nextInt();
        int K = sc.nextInt();
        int[] coins = new int[N];
        for (int i = 0; i < N; i++) {
            coins[i] = sc.nextInt();
        }
        long[] dp = new long[K + 1];

        for (int coin : coins) {
            if (coin > K) {
                continue;
            }

            dp[coin] += 1;
            for (int i = coin; i <= K; i++) {
                dp[i] += dp[i - coin];
            }
        }

        System.out.println(dp[K]);
    }
}